src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+static Oid GetRTEToSubqueryConverterReferenceRelId(
+	RTEToSubqueryConverterReference *rteToSubqueryConverterReference);
+static bool ShouldConvertLocalTableJoinsToSubqueries(List *rangeTableList, Oid
+													 resultRelationId);
+static bool HasUniqueFilter(RangeTblEntry *distRTE, List *distRTERestrictionList,
+							List *requiredAttrNumbersForDistRTE);
+static bool ShouldConvertDistributedTable(FromExpr *joinTree,
+										  RTEToSubqueryConverterReference *distRTEContext);
+static List * RequiredAttrNumbersForRelation(RangeTblEntry *relationRte,
+											 RecursivePlanningContext *planningContext);
+static RTEToSubqueryConverterContext * CreateRTEToSubqueryConverterContext(
+	RecursivePlanningContext *context,
+	List *
+	rangeTableList);
+static void GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexes);
+static RTEToSubqueryConverterReference * GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+																	   RTEToSubqueryConverterContext
+																	   *
+																	   rteToSubqueryConverterContext,
+																	   PlannerRestrictionContext
+																	   *
+																	   plannerRestrictionContext,
+																	   Oid
+																	   resultRelationId);
+static void GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+											 List **joinRangeTableEntries);
+static void RemoveFromRTEToSubqueryConverterContext(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext, Oid relationId);
+static bool FillLocalAndDistributedRTECandidates(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+	RTEToSubqueryConverterReference **
+	localRTECandidate,
+	RTEToSubqueryConverterReference **
+	distributedRTECandidate);
+
+/*
+ * ConvertUnplannableTableJoinsToSubqueries gets a query and the planner
+ * restrictions. As long as the query is not plannable by router planner,
+ * it converts either a local or distributed table to a subquery.
+ */
+void
+ConvertUnplannableTableJoinsToSubqueries(Query *query,
+										 RecursivePlanningContext *context)
+{
+	List *rangeTableList = NIL;
+	GetRangeTableEntriesFromJoinTree((Node *) query->jointree, query->rtable,
+									 &rangeTableList);
+	Oid resultRelationId = GetResultRelationId(query);
+	if (!ShouldConvertLocalTableJoinsToSubqueries(rangeTableList, resultRelationId))
+	{
+		return;
+	}
+
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext =
+		CreateRTEToSubqueryConverterContext(
+			context, rangeTableList);
+
+	RTEToSubqueryConverterReference *rteToSubqueryConverterReference =
+		GetNextRTEToConvertToSubquery(query->jointree, rteToSubqueryConverterContext,
+									  context->plannerRestrictionContext,
+									  resultRelationId);
+
+	PlannerRestrictionContext *plannerRestrictionContext =
+		FilterPlannerRestrictionForQuery(context->plannerRestrictionContext, query);
+	while (rteToSubqueryConverterReference && !IsRouterPlannable(query,
+																 plannerRestrictionContext))
```

Comment:
Why does it need to be router plannable?

What happens to a query like:
```sql
SELECT * FROM local_countries l JOIN distributed d ON (d.country_id = l.id) WHERE l.continent = 'Europe';
```

Would it just keep replacing tables with subqueries, such that the whole distributed table is pulled to the coordinator?

Reply:
I guess that's handled by GetNextRTEToConvertToSubquery